### ISSUE TITLE
Allow navigations and FK properties on non-collection complex types 

### ISSUE THREAD

**Issue by AndriySvyryd** [dotnet/efcore#31245]
The relationship itself will still be defined between two entity types.

Depends on #31236

**Comment by roji** [comment]
Adding needs-design as I believe we should allow (outgoing) navigations from complex types even when nested inside a collection.

**Comment by vgallegob** [comment]
Hello team,

I also need this feature. I am starting to adopt Complex Types in my domain, and I find it limiting not beeing able to have FKs inside of them. 

Here's a very simple example;

```
public abstract class UserActivityEntity : UserEntity
{
    public DateTime CreatedAt { get; set; }
    public required string CreatedByUserId { get; set; }
    public virtual AppUser? CreatedByUserNav { get; set; }

    public DateTime UpdatedAt { get; set; }
    public required string UpdatedByUserId { get; set; }
    public virtual AppUser? UpdatedByUserNav { get; set; }
}
```

could become 

```
public class UserActivityCT
{
    public DateTime At { get; set; }
    public required string ByUserId { get; set; }
    public virtual AppUser? ByUserNav { get; set; }
}

public abstract class UserActivityEntity : UserEntity
{
    public UserActivityCT Created { get; set; }
    public UserActivityCT Updated { get; set; }
}
```

But currently I get this error:

`Unable to create a 'DbContext' of type ''. The exception 'The expression 'l => l.Created.ByUserNav' is not a valid member access expression. The expression should represent a simple property or field access: 't => t.MyProperty'. (Parameter 'memberAccessExpression')' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728`

And so on and so on...

This is a small example but you can see the potential benefits of it. At the end of the day, it's just another column in the DataBase, so technicallities aside it would make sense to be able to do that, I think.

Is there any work around you can think of? Is this planned for .NET 9?

**Comment by roji** [comment]
@vgallegob you should already be able to use owned entities to map the above; this issue is specifically about allowing navigations on complex types, which are a new mapping strategy (we don't yet know if we'll be able to deliver this for 9.0). If you're having trouble with owned entities, please open a new issue with a runnable, minimal code sample.

**Comment by vgallegob** [comment]
I am already using Complex Types in my app. Would be nice to have it unified in Complex Types. For now I'll do this ones that have a FK with owned entities.

**Comment by CesarD** [comment]
This deserves more upvotes!
I want to leverage ComplexTypes for Value Objects in my projects but not possible since I cannot do something as simple as having an Address VO that holds a reference to a Country entity and its corresponding CountryId.

**Comment by adrielairaldo** [comment]
@vgallegob Beautiful use case you have put forward!

I just wanted to ask you why you said:

> At the end of the day, it's just another column in the DataBase

Your Value Object UserActivityCT will already be represented by two columns in your UserActivityEntity (one for At and one for ByUserId), so, I understand that what we want is to generate a FK from Created.ByUserId to User.Id. If I understand your case correctly, no new columns are needed.

In fact, a syntax like the one that already exists would be great (but it doesn't work):

`builder.HasOne<User>().WithMany().HasForeignKey(uae => uae.Created.ByUserId).IsRequired();` where (uae is UserActivityEntity)

Finally, my analysis must also be applied to the Updated field.

**Comment by AndriySvyryd** [comment]
@CesarD Why do you need to map `Address` as a complex type instead of an owned type?

**Comment by CesarD** [comment]
@AndriySvyryd because of the features ComplexTypes provide? Like being able to really use Records instead of mimicking them with classes and prevent having to track them as entities as it's the case of Owned Entities.

**Comment by AndriySvyryd** [comment]
Ok, then https://github.com/dotnet/efcore/issues/11457 could be an alternative way of supporting your scenario.

**Comment by CesarD** [comment]
Not sure about that. I'm looking to have support more closely to Value Objects rather than yet another type of entity.
Address would be a VO and because of that, it shouldn't be tracked like an entity.

**Comment by vgallegob** [comment]
All we want is a unified developer experience that allows mapping Value Objects in the DB. Navigation properties need to be represented in this unified way too, for convenience.

**Comment by c5racing** [comment]
Is there any movement on this with v10?

**Comment by roji** [comment]
This is in the backlog (see milestone); we do plan a big push on complex type support, but I think it's unlikely the specific feature will make it for 10.

**Comment by CesarD** [comment]
> This is in the backlog (see milestone); we do plan a big push on complex type support, but I think it's unlikely the specific feature will make it for 10.

This is a huge deal breaker for implementing Value Objects in a consistent manner if we can't fully leverage complex types.

Is it a capacity problem or just other priorities?

**Comment by roji** [comment]
@CesarD I agree it's something we should implement, but both priorities and capacity will probably mean that it doesn't make it into the release. I'll also point out that this issue has around 40 votes, while other complex type-related issues have much, much more, so not everyone seems to agree that it's a huge deal breaker.

**Comment by CesarD** [comment]
Well, by my count it makes it into the top 10 of issues tagged with `area-complex-types` that have the most reactions 👍🏻

I figured the big push you mentioned might include the top 10 ones? 🙂

**Comment by roji** [comment]
> I figured the big push you mentioned might include the top 10 ones?

Probably not. As I wrote, overall this is quite low in the overall list of issues with only 46 votes.

**Comment by AndriySvyryd** [comment]
**Design notes:** Complex types don't have identity, so there couldn't ever be an inverse "navigation" as it would be impossible to populate in a meaningful way. But also, it would be prohibitively expensive and error-prone to fix up the navigations on the complex types as the related entity types get added and removed and the containing complex collection is modified because the internal representation of backreferences would need to be implemented in terms of paths. If any part of a path is changed then the cost of updating the internal data structure would be O(C*R) where C is the number of complex type instances with a navigation and R is the number of tracked entities of the navigation target type.

**Comment by roji** [comment]
@AndriySvyryd if/when we get around to it, let's try to think together about possible alternative fix-up designs which would make this feasible. I get that our current fixup design makes this problematic, but this doesn't feel like it should be an unsolvable problem (even though switching fixup designs itself could be an effort that ultimately isn't worth the benefit here).

**Comment by AndriySvyryd** [comment]
As an alternative, https://github.com/dotnet/efcore/issues/31247 would enable injecting a `DbContext` instance in the complex type, this would allow to lookup the related entity in the implementation of a pseudo-navigation.
And use https://github.com/dotnet/efcore/issues/33171 for query support.

**Comment by vgallegob** [comment]
Do you mean something like

```
public class UserActivityCT
{
    public DateTime At { get; set; }
    public required string ByUserId { get; set; }
    public virtual AppUser? ByUserNav => AppDbContext.Users.Find(ByUserId);
    private virtual AppDbContext DbContext { get; set; }
}
```

This would require Sync evaluation of the query wouldn't it?

**Comment by AndriySvyryd** [comment]
> This would require Sync evaluation of the query wouldn't it?

Yes, if it's not loaded yet. https://github.com/dotnet/efcore/issues/33171 would allow this to be eager loaded

**Comment by vgallegob** [comment]
I see the point. 

It would preload the data and allow us to access the "navigation property" but it would never be tracked by EF and we could never edit it.  Right?

So the main problem is the tracking extra types?

What if instead of tracking Complex Types, the navigation property was appended to the owning type?
Crazy idea, but, could EF core convert this:

```
public class UserActivityCT
{
    public DateTime At { get; set; }
    public required string ByUserId { get; set; }
    public virtual AppUser? ByUserNav { get; set; }
}

public abstract class UserActivityEntity : UserEntity
{
    public UserActivityCT Created { get; set; }
    public UserActivityCT Updated { get; set; }
}
```

to this internally:

```
public class UserActivityCT
{
    public DateTime At { get; set; }
    public required string ByUserId { get; set; }
    public virtual AppUser? ByUserNav { get; set; }
}

public abstract class UserActivityEntity : UserEntity
{
    public UserActivityCT Created { get; set; }
    public UserActivityCT Updated { get; set; }
    private virtual AppUser? Created_ByUserNav { get; set; }
    private virtual AppUser? Updated_ByUserNav { get; set; }
}
```

And track only the UserActivityEntity?
At the end of the day, we could flattern all navigation properties in the Entity.
That should be similar in perf since that is the exact number of navigation properties i have today in this example:

```
public abstract class UserActivityEntity : UserEntity
{
    public DateTime CreatedAt { get; set; }
    public required string CreatedByUserId { get; set; }
    public virtual AppUser? CreatedByUserNav { get; set; }

    public DateTime UpdatedAt { get; set; }
    public required string UpdatedByUserId { get; set; }
    public virtual AppUser? UpdatedByUserNav { get; set; }
}
```

I think the part where EF core assingns values from the Entity to the ComplexType is a bit complicated, but solvable perhaps?
Not sure. But this would allow editing the navigation property values, which is necesary in any case.

**Comment by AndriySvyryd** [comment]
> It would preload the data and allow us to access the "navigation property" but it would never be tracked by EF and we could never edit it. Right?

Right, but you can still change the FK property, which is what is actually persisted.

> Crazy idea, but, could EF core convert this:

That's essentially the approach implied in the first comment, but it wouldn't work for complex collections.
